syntax = "proto3";

package modelapi;

service ModelService {
    rpc ListModels (Empty) returns (ModelList);
    rpc TrainModel (TrainRequest) returns (TrainResponse);
    rpc Predict (PredictRequest) returns (PredictResponse);
    rpc DeleteModel (DeleteRequest) returns (DeleteResponse);
}

message Empty {}

message ModelList {
    map<string, ModelSpec> models = 1;
}

message ModelSpec {
    repeated string parameters = 1;
}

message TrainRequest {
    string type = 1;
    map<string, string> parameters = 2;
    repeated FeatureSet features = 3;
    repeated double targets = 4;
}

message FeatureSet {
    map<string, double> features = 1;
}

message TrainResponse {
    string status = 1;
    string model_id = 2;
}

message PredictRequest {
    string model_id = 1;
    repeated FeatureSet features = 2;
}

message PredictResponse {
    string model_id = 1;
    repeated double predictions = 2;
}

message DeleteRequest {
    string model_id = 1;
}

message DeleteResponse {
    string status = 1;
}
